src/Derivative/Parser.hs

```diff
@@ -39,104 +39,106 @@ import Data.Monoid hiding (Alt)
 sep1 s p = (:) <$> p <*> many (s *> p)
 
-sep :: Combinator v sep -> Combinator v a -> Combinator v [a]
+sep :: Combinator v t sep -> Combinator v t a -> Combinator v t [a]
 sep s p = s `sep1` p <|> pure []
 
 oneOf :: (Foldable t, Alternative f) => t (f a) -> f a
 oneOf = getAlt . foldMap Monoid.Alt
 
 infixl 4 `cat`
 
-cat :: Combinator v a -> Combinator v b -> Combinator v (a, b)
+cat :: Combinator v t a -> Combinator v t b -> Combinator v t (a, b)
 cat a = compact' . rec . Cat a
 
-char :: Char -> Combinator v Char
-char = rec . Chr
+char :: Char -> Combinator v Char Char
+char = rec . Sat . Equal
 
-category :: GeneralCategory -> Combinator v Char
-category = rec . Uni
+category :: GeneralCategory -> Combinator v Char Char
+category = rec . Sat . Category
 
-delta :: Combinator v a -> Combinator v a
+delta :: Combinator v t a -> Combinator v t a
 delta = compact' . rec . Del
 
-ret :: [a] -> Combinator v a
+ret :: [a] -> Combinator v t a
 ret = rec . Ret
 
 infixr 2 `label`
 
-label :: Combinator v a -> String -> Combinator v a
+label :: Combinator v t a -> String -> Combinator v t a
 label p = compact' . rec . Lab p
 
-string :: String -> Combinator v String
-string string = sequenceA (rec . Chr <$> string)
+string :: String -> Combinator v Char String
+string string = sequenceA (rec . Sat . Equal <$> string)
 
-mu :: (Combinator v a -> Combinator v a) -> Combinator v a
+mu :: (Combinator v t a -> Combinator v t a) -> Combinator v t a
 mu f = Graph.mu $ \ v -> case f (var v) of
   Rec (In r) -> r
   p -> p `Lab` ""
 
-parser :: (forall v. Combinator v a) -> Parser a
+parser :: (forall v. Combinator v t a) -> Parser t a
 parser r = compact $ Graph r
 
-combinator :: Parser a -> Combinator v a
+combinator :: Parser t a -> Combinator v t a
 combinator = unGraph
 
 
 -- Types
 
 -- | A parser type encoding concatenation, alternation, repetition, &c. as first-order constructors.
-data ParserF f a where
-  Cat :: f a -> f b -> ParserF f (a, b)
-  Alt :: f a -> f a -> ParserF f a
-  Rep :: f a -> ParserF f [a]
-  Map :: (a -> b) -> f a -> ParserF f b
-  Bnd :: f a -> (a -> f b) -> ParserF f b
-  Chr :: Char -> ParserF f Char
-  Uni :: GeneralCategory -> ParserF f Char
-  Ret :: [a] -> ParserF f a
-  Nul :: ParserF f a
-  Lab :: f a -> String -> ParserF f a
-  Del :: f a -> ParserF f a
-
-type Parser = Graph ParserF
-type Combinator v = Rec ParserF v
+data ParserF t f a where
+  Cat :: f a -> f b -> ParserF t f (a, b)
+  Alt :: f a -> f a -> ParserF t f a
+  Rep :: f a -> ParserF t f [a]
+  Map :: (a -> b) -> f a -> ParserF t f b
+  Bnd :: f a -> (a -> f b) -> ParserF t f b
+  Sat :: Predicate t -> ParserF t f t
+  Ret :: [a] -> ParserF t f a
+  Nul :: ParserF t f a
+  Lab :: f a -> String -> ParserF t f a
+  Del :: f a -> ParserF t f a
+
+type Parser t = Graph (ParserF t)
+type Combinator v t = Rec (ParserF t) v
+
+data Predicate t where
+  Equal :: Eq t => t -> Predicate t
+  Category :: GeneralCategory -> Predicate Char
```

Comment:
The `Predicate` type allows us to retain straightforward `Show` and `Eq` instances.